src/iface-opc-ua.c

```diff
@@ -0,0 +1,347 @@
+	UA_NodeId_init(parent);
+
+	rc = get_node_id(parent, data_item->namespace,
+			 data_item->identifier_type,
+			 data_item->identifier);
+	if (rc == UA_STATUSCODE_GOOD)
+		rc = UA_Client_readValueAttribute(opc_ua_client,
+						  *parent, value);
+
+	free(parent);
+	return rc;
+}
+
+static UA_StatusCode get_read_value(UA_Variant *value,
+				    uint8_t value_type,
+				    int value_type_size,
+				    union opc_ua_types *tmp)
+{
+	UA_StatusCode rc = UA_STATUSCODE_GOOD;
+
+	switch (value_type) {
+	case KNOT_VALUE_TYPE_BOOL:
+		if (value_type_size == 1)
+			tmp->val_bool = *(UA_Boolean *)value->data;
+		else if (value_type_size == 8)
+			tmp->val_bool = *(UA_SByte *)value->data;
+		else
+			rc = UA_STATUSCODE_BAD;
+		break;
+	case KNOT_VALUE_TYPE_FLOAT:
+		tmp->val_float = *(UA_Float *)value->data;
+		break;
+	case KNOT_VALUE_TYPE_INT:
+		if (value_type_size == 8)
+			tmp->val_i8 = *(UA_Byte *)value->data;
+		else if (value_type_size == 16)
+			tmp->val_i16 = *(UA_Int16 *)value->data;
+		else if (value_type_size == 32)
+			tmp->val_i32 = *(UA_Int32 *)value->data;
+		else
+			rc = UA_STATUSCODE_BAD;
+		break;
+	case KNOT_VALUE_TYPE_UINT:
+		if (value_type_size == 8)
+			tmp->val_u8 = *(UA_SByte *)value->data;
+		if (value_type_size == 16)
+			tmp->val_u16 = *(UA_UInt16 *)value->data;
+		else if (value_type_size == 32)
+			tmp->val_u32 = *(UA_UInt32 *)value->data;
+		else
+			rc = UA_STATUSCODE_BAD;
+		break;
+	case KNOT_VALUE_TYPE_INT64:
+		tmp->val_i64 = *(UA_Int64 *)value->data;
+		break;
+	case KNOT_VALUE_TYPE_UINT64:
+		tmp->val_u64 = *(UA_UInt64 *)value->data;
+		break;
+	default:
+		rc = UA_STATUSCODE_BAD;
+	}
+
+	return rc;
+}
+
+int iface_opc_ua_config(struct knot_data_item *data_item,
+			       struct knot_thing *thing)
+{
+	return 0;
+}
+
+int iface_opc_ua_read_data(struct knot_data_item *data_item)
+{
+	UA_StatusCode rc = UA_STATUSCODE_GOOD;
+	union opc_ua_types tmp;
+	UA_Variant *value = UA_Variant_new();
+
+	rc = setup_read_data(value, data_item);
+	memset(&tmp, 0, sizeof(tmp));
+
+	if (rc == UA_STATUSCODE_GOOD) {
+		rc = get_read_value(value, data_item->schema.value_type,
+				    data_item->value_type_size, &tmp);
+	}
+
+	if (rc != UA_STATUSCODE_GOOD) {
+		l_error("Failed to read from OPC UA: %#010x",
+			rc);
+		l_io_destroy(opc_ua_io);
+		opc_ua_io = NULL;
+	} else {
+		memcpy(&data_item->current_val, &tmp, sizeof(tmp));
+	}
+
+	UA_Variant_delete(value);
+	return -rc;
+}
+
+static void on_disconnected(struct l_io *io, void *user_data)
+{
+	if (disconn_cb)
+		disconn_cb(user_data);
+
+	if (connect_to)
+		l_timeout_modify(connect_to, RECONNECT_TIMEOUT);
+}
+
+static void attempt_connect(struct l_timeout *to, void *user_data)
+{
+	UA_StatusCode rc = UA_STATUSCODE_GOOD;
+
+	l_debug("Trying to connect to OPC UA Server");
+
+	/* Check and destroy if an IO is already allocated */
+	if (opc_ua_io) {
+		l_io_destroy(opc_ua_io);
+		opc_ua_io = NULL;
+	}
+
+	if (strlen(thing_opc_ua.driver_login) != STRING_EMPTY) {
```

Comment:
The connection is made with or without authentication depending on the configuration, right? If so, what do you think of making the intention of this code snippet a little clearer? For example, creating a constant that identifies whether the connection is authenticated or not.

```
const int is_authenticated_connection = (strlen(thing_opc_ua.driver_login) != STRING_EMPTY) ? 1 : 0;

if (is_authenticated_connection){
…
} else {
}
```